x-pack/plugins/fleet/server/tasks/unenroll_inactive_agents_task.ts

```diff
@@ -0,0 +1,191 @@
+import { agentPolicyService } from '../services';
+
+export const TYPE = 'fleet:unenroll-inactive-agents-task';
+export const VERSION = '1.0.0';
+const TITLE = 'Fleet Deleted Files Periodic Tasks';
+const SCOPE = ['fleet'];
+const INTERVAL = '10m';
+const TIMEOUT = '1m';
+const POLICIES_BATCHSIZE = 100;
+const UNENROLLMENT_BATCHSIZE = 1000;
+
+interface UnenrollInactiveAgentsTaskSetupContract {
+  core: CoreSetup;
+  taskManager: TaskManagerSetupContract;
+  logFactory: LoggerFactory;
+}
+
+interface UnenrollInactiveAgentsTaskStartContract {
+  taskManager: TaskManagerStartContract;
+}
+
+export class UnenrollInactiveAgentsTask {
+  private logger: Logger;
+  private wasStarted: boolean = false;
+  private abortController = new AbortController();
+
+  constructor(setupContract: UnenrollInactiveAgentsTaskSetupContract) {
+    const { core, taskManager, logFactory } = setupContract;
+    this.logger = logFactory.get(this.taskId);
+
+    taskManager.registerTaskDefinitions({
+      [TYPE]: {
+        title: TITLE,
+        timeout: TIMEOUT,
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          return {
+            run: async () => {
+              return this.runTask(taskInstance, core);
+            },
+            cancel: async () => {
+              this.abortController.abort('Task timed out');
+            },
+          };
+        },
+      },
+    });
+  }
+
+  public start = async ({ taskManager }: UnenrollInactiveAgentsTaskStartContract) => {
+    if (!taskManager) {
+      this.logger.error('[UnenrollInactiveAgentsTask] Missing required service during start');
+      return;
+    }
+
+    this.wasStarted = true;
+    this.logger.info(`[UnenrollInactiveAgentsTask] Started with interval of [${INTERVAL}]`);
+
+    try {
+      await taskManager.ensureScheduled({
+        id: this.taskId,
+        taskType: TYPE,
+        scope: SCOPE,
+        schedule: {
+          interval: INTERVAL,
+        },
+        state: {},
+        params: { version: VERSION },
+      });
+    } catch (e) {
+      this.logger.error(`Error scheduling task UnenrollInactiveAgentsTask, error: ${e.message}`, e);
+    }
+  };
+
+  private get taskId(): string {
+    return `${TYPE}:${VERSION}`;
+  }
+
+  private endRun(msg: string = '') {
+    this.logger.info(`[UnenrollInactiveAgentsTask] runTask ended${msg ? ': ' + msg : ''}`);
+  }
+
+  public async unenrollInactiveAgents(
+    esClient: ElasticsearchClient,
+    soClient: SavedObjectsClientContract
+  ) {
+    this.logger.debug(
+      `[UnenrollInactiveAgentsTask] Fetching agent policies with unenroll_timeout > 0`
+    );
+    // find all agent policies that are not managed and having unenroll_timeout > 0
+    // limit batch size to 100 to avoid scale issues
+    const policiesKuery = `${AGENT_POLICY_SAVED_OBJECT_TYPE}.is_managed: false AND ${AGENT_POLICY_SAVED_OBJECT_TYPE}.unenroll_timeout > 0`;
```

Comment:
I suppose if `unenroll_timeout` is not set on an agent policy, it is not going to show up in the results?

Reply:
Yes I checked it by setting `unenroll_timeout` on a policy and then trying this query. For instance from the searchbox in the agent policies page it filters out correctly. I had 4 policies and set the unenroll_timeout only on `Policy with unenroll_timeout`:
![Screenshot 2024-08-06 at 12 04 38](https://github.com/user-attachments/assets/9f76ffa9-3a95-44bb-aeb0-e3c69c964e17)

![Screenshot 2024-08-06 at 11 54 11](https://github.com/user-attachments/assets/d29ad645-46db-4dc0-af4a-1e3c60116910)
![Screenshot 2024-07-30 at 11 08 31](https://github.com/user-attachments/assets/c3f3ea2c-7cd8-49b7-a957-b5072902ad78)